app/services/csv_import/companies.rb

```diff
@@ -0,0 +1,51 @@
+module CSVImport
+  class Companies < BaseImporter
+    include UploaderHelpers
+
+    def import
+      import_each_csv_row(csv) do |row|
+        company = prepare_company(row)
+
+        company.assign_attributes(company_attributes(row))
+
+        was_new_record = company.new_record?
+        any_changes = company.changed?
+
+        company.save!
+
+        update_import_results(was_new_record, any_changes)
+      end
+    end
+
+    private
+
+    def resource_klass
+      Company
+    end
+
+    def prepare_company(row)
+      find_record_by(:id, row) ||
+        find_record_by(:isin, row) ||
+        find_record_by(:name, row) ||
+        resource_klass.new
+    end
+
+    # Builds resource attributes from a CSV row.
+    #
+    # If not present in DB, will create related:
+    # - sector
+    #
+    def company_attributes(row)
+      {
+        name: row[:name],
+        isin: row[:isin],
+        sector: Sector.find_or_create_by!(name: row[:sector]),
```

Comment:
I would try to find by downcased name, to prevent creating multiple sectors with the same name.

```
    def find_or_create_sector(row)
      return unless row[:sector].present?

      Sector.where('lower(name) = ?', row[:sector].downcase).first ||
        Sector.new(name: row[:sector])
    end
```

Reply:
updated